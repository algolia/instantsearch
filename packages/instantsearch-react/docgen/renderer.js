import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
