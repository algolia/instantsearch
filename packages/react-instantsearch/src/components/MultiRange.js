import React, {PropTypes, Component} from 'react';
import List from './List';
import classNames from './classNames.js';

const cx = classNames('MultiRange');

class MultiRange extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.node.isRequired,
      value: PropTypes.string.isRequired,
    })).isRequired,
    refine: PropTypes.func.isRequired,
    transformItems: PropTypes.func,
  };

  renderItem = item => {
    const {refine} = this.props;

    return (
      <label>
        <input
          {...cx('itemRadio', item.isRefined && 'itemRadioSelected')}
          type="radio"
          checked={item.isRefined}
          onChange={refine.bind(null, item.value)}
        />
        <span {...cx('itemBox', 'itemBox', item.isRefined && 'itemBoxSelected')}></span>
        <span {...cx('itemLabel', 'itemLabel', item.isRefined && 'itemLabelSelected')}>
          {item.label}
        </span>
      </label>
    );
  };

  render() {
    const {items} = this.props;

    return (
      <List
        renderItem={this.renderItem}
        showMore={false}
        cx={cx}
        items={items.map(item => ({...item, key: item.value}))}
      />
    );
  }
}

export default MultiRange;
