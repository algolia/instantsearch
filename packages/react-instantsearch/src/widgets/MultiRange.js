import connectMultiRange from '../connectors/connectMultiRange.js';
import MultiRangeComponent from '../components/MultiRange.js';

/**
 * MultiRange is a widget used for selecting the range value of a numeric attribute.
 * @name MultiRange
 * @kind component
 * @category widget
 * @propType {string} attributeName - the name of the attribute in the records
 * @propType {{label: string, start: number, end: number}[]} items - List of options. With a text label, and upper and lower bounds.
 * @propType {string} defaultRefinement - the value of the item selected by default, follow the format "min:max".
 * @themeKey ais-MultiRange__root - The root component of the widget
 * @themeKey ais-MultiRange__items - The container of the items
 * @themeKey ais-MultiRange__item - A single item
 * @themeKey ais-MultiRange__itemSelected - The selected item
 * @themeKey ais-MultiRange__itemLabel - The label of an item
 * @themeKey ais-MultiRange__itemLabelSelected - The selected label item
 * @themeKey ais-MultiRange__itemRadio - The radio of an item
 * @themeKey ais-MultiRange__itemRadioSelecte - The selected radio item
 * @example
 * import React from 'react';
 *
 * import {InstantSearch, MultiRange} from '../packages/react-instantsearch/dom';
 *
 * export default function App() {
 *   return (
 *     <InstantSearch
 *       className="container-fluid"
 *       appId="latency"
 *       apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *       indexName="ikea"
 *     >
 *       <MultiRange attributeName="price"
 *           items={[
 *             {end: 10, label: '<$10'},
 *             {start: 10, end: 100, label: '$10-$100'},
 *             {start: 100, end: 500, label: '$100-$500'},
 *             {start: 500, label: '>$500'},
 *           ]}
 *        />
 *     </InstantSearch>
 *   );
 * }
 */
export default connectMultiRange(MultiRangeComponent);
