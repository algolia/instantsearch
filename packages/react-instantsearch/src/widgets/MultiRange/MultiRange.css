.root {
}

.items {
}

.item {
}

.itemSelected {
}

.itemLabel {
}

.itemRadio {
}