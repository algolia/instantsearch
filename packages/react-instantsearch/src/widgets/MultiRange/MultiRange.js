import React, {PropTypes, Component} from 'react';

import themeable from '../../core/themeable';

import List from '../../components/List';

import theme from './MultiRange.css';

class MultiRange extends Component {
  static propTypes = {
    applyTheme: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.node.isRequired,
      value: PropTypes.string.isRequired,
    })).isRequired,
    selectedItem: PropTypes.string.isRequired,
    refine: PropTypes.func.isRequired,
  };

  renderItem = (item, selected) => {
    const {applyTheme, refine} = this.props;

    return (
      <label>
        <input
          {...applyTheme('itemRadio', 'itemRadio')}
          type="radio"
          checked={selected}
          onChange={refine.bind(null, item.value)}
        />
        <span {...applyTheme('itemLabel', 'itemLabel')}>
          {item.label}
        </span>
      </label>
    );
  };

  render() {
    const {items, selectedItem, applyTheme} = this.props;

    return (
      <List
        renderItem={this.renderItem}
        showMore={false}
        applyTheme={applyTheme}
        items={items}
        selectedItems={[selectedItem]}
      />
    );
  }
}

export default themeable(theme)(MultiRange);
