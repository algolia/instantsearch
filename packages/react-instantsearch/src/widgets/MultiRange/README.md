---
title: MultiRange
layout: widget.pug
nav_groups:
  - widgets
---

# MultiRange

Lets the user pick between a list of refinement ranges for a particular numeric facet.

## Props

<!-- props default ./index.js -->

### Theme

`root`, `items`, `item`, `itemSelected`, `itemLabel`, `itemLabelSelected`, `itemRadio`, `itemRadioSelected`

## Implementing your own MultiRange

See [Making your own widgets](../Customization.md) for more information on how to use the `MultiRange.connect` HOC.

```
import {MultiRange.connect} from 'react-instantsearch';

function Item(props) {
  const value = props.selected ? null : props.item.value;
  return (
    <a
      key={item.value}
      className={props.selected ? 'selected' : 'not-selected'}
      href={props.createURL(value)}
      onClick={e => {
        e.preventDefault();
        props.refine(value);
      }}
    >
      {item.label}
    </a>
  );
}

function MyMultiRange(props) {
  return (
    <div>
      {props.items.map(item =>
        <Item
          key={item.value}
          item={item}
          selected={item.value === props.selectedItem}
          refine={props.refine}
          createURL={props.createURL}
        />
      )}
    </div>
  );
}

// `MultiRange.connect` accepts the same `id`, `attributeName` and
// `items` props as `MultiRange`.
// `items` are transformed into a simpler `{label: ?node, value: string}`
// shape before being passed to the connected component.
export default MultiRange.connect(MyMultiRange);
```
