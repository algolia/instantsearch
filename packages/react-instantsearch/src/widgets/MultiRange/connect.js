import {PropTypes} from 'react';
import {find} from 'lodash';

import createConnector from '../../core/createConnector';

function stringifyItem(item) {
  if (typeof item.start === 'undefined' && typeof item.end === 'undefined') {
    return '';
  }
  return `${item.start ? item.start : ''}:${item.end ? item.end : ''}`;
}

function parseItem(value) {
  if (value.length === 0) {
    return {start: null, end: null};
  }
  const [startStr, endStr] = value.split(':');
  return {
    start: startStr.length > 0 ? parseInt(startStr, 10) : null,
    end: endStr.length > 0 ? parseInt(endStr, 10) : null,
  };
}

function getId(props) {
  return props.id || props.attributeName;
}

function getSelectedItem(props, state) {
  const id = getId(props);
  if (typeof state[id] !== 'undefined') {
    return state[id];
  }
  if (props.defaultSelectedItem) {
    return props.defaultSelectedItem;
  }
  return '';
}

export default createConnector({
  displayName: 'AlgoliaMultiRange',

  propTypes: {
    /**
     * URL state serialization key. Defaults to the value of `attributeName`.
     * The state of this widget takes the shape of a `string` with a pattern of
     * `'{start}:{end}'` which corresponds to the current selected item.
     * For instance, when the selected item is `{start: 10, end: 20}`, the state
     * of the widget is `'10:20'`. When `start` isn't defined, the state of the
     * widget is `':{end}'`, and the same way around when `end` isn't defined.
     * However, when neither `start` nor `end` are defined, the state is an
     * empty string.
     * @public
     */
    id: PropTypes.string,

    /**
     * Name of the attribute for faceting
     * @public
     */
    attributeName: PropTypes.string.isRequired,

    /**
     * List of options.
     * @public
     * @defines MultiRangeItem
     */
    items: PropTypes.arrayOf(PropTypes.shape({
      /**
       * Node to render in place of the option.
       */
      label: PropTypes.node,
      /**
       * Inclusive start of the range.
       */
      start: PropTypes.number,
      /**
       * Inclusive end of the range.
       */
      end: PropTypes.number,
    })).isRequired,
  },

  getProps(props, state) {
    const {items} = props;
    const selectedItem = getSelectedItem(props, state);

    return {
      items: items.map(item => ({
        label: item.label,
        value: stringifyItem(item),
      })),
      selectedItem,
    };
  },

  refine(props, state, nextSelectedItem) {
    return {
      ...state,
      [getId(props, state)]: nextSelectedItem,
    };
  },

  getSearchParameters(searchParameters, props, state) {
    const {attributeName} = props;
    const {start, end} = parseItem(getSelectedItem(props, state));

    if (start) {
      searchParameters = searchParameters.addNumericRefinement(
        attributeName,
        '>=',
        start
      );
    }
    if (end) {
      searchParameters = searchParameters.addNumericRefinement(
        attributeName,
        '<=',
        end
      );
    }
    return searchParameters;
  },

  getMetadata(props, state) {
    const id = getId(props);
    const value = getSelectedItem(props, state);
    const filters = [];
    if (value !== '') {
      const {label} = find(props.items, item => stringifyItem(item) === value);
      filters.push({
        key: `${id}.${value}`,
        label: `${props.attributeName}: ${label}`,
        clear: nextState => ({
          ...nextState,
          [id]: '',
        }),
      });
    }
    return {id, filters};
  },
});
