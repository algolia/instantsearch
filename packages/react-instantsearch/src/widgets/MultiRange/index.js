import connect from './connect';
import MultiRange from './MultiRange';

const Connected = connect(MultiRange);
Connected.connect = connect;
export default Connected;
