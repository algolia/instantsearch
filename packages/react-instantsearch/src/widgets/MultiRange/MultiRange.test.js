/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import MultiRange from './MultiRange';

describe('RangeInput', () => {
  it('supports passing items values', () => {
    const tree = renderer.create(
      <MultiRange
        createURL={() => '#'}
        refine={() => null}
        items={[
          {label: 'label', value: '10:'},
          {label: 'label', value: '10:20'},
          {label: 'label', value: '20:30'},
          {label: 'label', value: '30:'},
        ]}
        selectedItem=""
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('supports having a selected item', () => {
    const tree = renderer.create(
      <MultiRange
        createURL={() => '#'}
        refine={() => null}
        items={[
          {label: 'label', value: '10:'},
          {label: 'label', value: '10:20'},
          {label: 'label', value: '20:30'},
          {label: 'label', value: '30:'},
        ]}
        selectedItem="10:20"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
