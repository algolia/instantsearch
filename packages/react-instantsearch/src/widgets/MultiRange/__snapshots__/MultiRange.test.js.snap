exports[`RangeInput supports having a selected item 1`] = `
<div
  className="root">
  <div
    className="items">
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
    <div
      className="item itemSelected undefined undefined">
      <label>
        <input
          checked={true}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`RangeInput supports passing items values 1`] = `
<div
  className="root">
  <div
    className="items">
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
    <div
      className="item false undefined undefined">
      <label>
        <input
          checked={false}
          className="itemRadio"
          onChange={[Function bound refine]}
          type="radio" />
        <span
          className="itemLabel">
          label
        </span>
      </label>
    </div>
  </div>
</div>
`;
